# Linear Regression Using Scikit Learn

We will be learning the basics of Machine Learning and Linear Regression Algorithms.

What is **linear regression**?

Linear regression analysis is used to predict the value of a variable based on the value of another variable. The variable you want to predict is called the dependent variable. The variable you are using to predict the other variable's value is called the independent variable.

In this notebook we will cover:
* A typical problem statement for machine learning
* Downloading and exploring a dataset for machine learning
* Linear regression with one variable using Scikit-learn
* Linear regression with multiple variables
* Using categorical features for machine learning
* Regression coefficients and feature importance
* Other models and techniques for regression using Scikit-learn

**QUESTION**: 

ACME Insurance Inc. offers affordable health insurance to thousands of customer all over the United States. As the lead data scientist at ACME, you're tasked with creating an automated system to estimate the annual medical expenditure for new customers, using information such as their age, sex, BMI, children, smoking habits and region of residence.

## Importing the Data

In [2]:
dataset_url = 'https://raw.githubusercontent.com/JovianML/opendatasets/master/data/medical-charges.csv'

In [3]:
from urllib.request import urlretrieve

urlretrieve(dataset_url, 'medical.csv')

('medical.csv', <http.client.HTTPMessage at 0x7fe3a9374940>)

In [4]:
import pandas as pd

medical_df = pd.read_csv('medical.csv')

In [5]:
medical_df.head()

age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
1   18    male  33.770         1     no  southeast   1725.55230
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520

In [6]:
medical_df.shape

(1338, 7)

The dataframe has 1338 rows and 7 columns. The one we need to predict is the 'charges', given the other column values for a new patient.

In [8]:
#Checking the data types of each features
medical_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1338 non-null   int64  
 1   sex       1338 non-null   object 
 2   bmi       1338 non-null   float64
 3   children  1338 non-null   int64  
 4   smoker    1338 non-null   object 
 5   region    1338 non-null   object 
 6   charges   1338 non-null   float64
dtypes: float64(2), int64(2), object(3)
memory usage: 73.3+ KB


There are no null values in the dataframe. Therefore no need for imputation of values. 

Here age, bmi, children, charges are integers (numerical values), whereas sex, smoker, region are string (categorical values)

In [9]:
#check statistics of numerical column
medical_df.describe()

age          bmi     children       charges
count  1338.000000  1338.000000  1338.000000   1338.000000
mean     39.207025    30.663397     1.094918  13270.422265
std      14.049960     6.098187     1.205493  12110.011237
min      18.000000    15.960000     0.000000   1121.873900
25%      27.000000    26.296250     0.000000   4740.287150
50%      39.000000    30.400000     1.000000   9382.033000
75%      51.000000    34.693750     2.000000  16639.912515
max      64.000000    53.130000     5.000000  63770.428010

Seems like the charges are skewed to the right skewed as the maximum of the charges is far above the median. Almost, 75% of the population has 2 children.

**If your BMI is less than 18.5, it falls within the underweight range. If your BMI is 18.5 to 24.9, it falls within the Healthy Weight range. If your BMI is 25.0 to 29.9, it falls within the overweight range. If your BMI is 30.0 or higher, it falls within the obese range.**

Average of the population seems to be either Overweight or Obese, and only very few are underweighted. Only >25% of the population falls under healthy weight range.
